src/js/addons/widgets/Widget.js

```diff
@@ -1,34 +1,62 @@
+/**
+ *  A widget is a standalone unit of functionality for a
+ *  given matching context in Tangram YAML.
+ *
+ *  An instance of the widget class stores a reference to
+ *  its DOM element (and children) and can be extended
+ *  with any additional functionality.
+ *
+ */
+
+import { getPosition } from '../../core/common.js';
+
+// Store reference to the editor on this module
+let editor;
+
 export default class Widget {
-    constructor (datum) {
-        const matchTypes = [
-            'value',
-            'key',
-            'address',
-        ];
-
-        for (const key of matchTypes) {
-            if (datum[key]) {
-                this.matchAgainst = key;
-                this.matchPattern = datum[key];
-                break;
-            }
-        }
+    constructor (def, key, cm) {
+        this.key = key;
+        this.definition = def;
+        this.el = this.createEl();
+
+        // Store a reference to the editor
+        editor = cm;
     }
 
-    match (keyPair) {
-        if (this.matchAgainst) {
-            return RegExp(this.matchPattern).test(keyPair[this.matchAgainst]);
-        }
-        else {
-            return false;
+    // Returns a widget element.
+    // This is a simple bare-bones example.
+    // Widgets that extend from this should create
+    // their own DOM and avoid calling super()
+    createEl () {
+        return document.createDocumentFragment();
+    }
+
+    // Removes the widget element from DOM.
+    destroyEl () {
+        if (this.el && this.el.parentNode) {
+            this.el.parentNode.removeChild(this.el);
         }
     }
 
-    wrap (dom, keyPair) {
-        return {
-            dom,
-            line: keyPair.pos.line,
-            index: keyPair.index
-        };
+    // Returns x, y position of the upper left corner
+    // of the DOM element for the widget. Useful
+    // for positioning modals, etc.
+    getPosition () {
+        return getPosition(this.el);
+    }
+
+    // Future
```

Comment:
I'm excited for this. Improving the way we update Widget positions will be great